File: kotlin/Dynamic-Programming/Easy/819.最常见的单词.kt
/*
 * @lc app=leetcode.cn id=819 lang=kotlin
 *
 * [819] 最常见的单词
 */

// @lc code=start
class Solution {
    fun mostCommonWord(paragraph: String, banned: Array<String>): String { // 自己写的
        val set = HashSet<String>();
        for (word in banned) {
            set.add(word);
        }
        val map = HashMap<String, Int>();
        var maxWords = "";
        var maxCount = 0;
        var words = "";
        for (i in paragraph) {
            if (Character.isLetter(i)) {
                words += i.lowercase();
            } else if (words != "") {
                if (set.contains(words)) {
                    words = "";
                    continue;
                }

                if (map.containsKey(words)) {
                    val max = map.get(words)!! + 1;
                    map.put(words, max);
                    if (max > maxCount) {
                        maxWords = words;
                        maxCount = max;
                    }
                } else {
                    if (maxCount == 0) {
                        maxWords = words;
                        maxCount = 1;
                    }
                    map.put(words, 1);
                }
                words = "";
            }
        }
        if (maxCount == 0) {
            maxWords = words.lowercase();
        }
        return maxWords;
    }
}
// @lc code=end

File: jsTest/dynamic-programming/easy/819.最常见的单词.js
/*
 * @lc app=leetcode.cn id=819 lang=javascript
 *
 * [819] 最常见的单词
 */
// @lc code=start
/**
 * @param {string} paragraph
 * @param {string[]} banned
 * @return {string}
 */
var mostCommonWord = function(paragraph, banned) {
    const isLetter = (ch) => {
        return (ch >= "a" && ch <= "z") || (ch >= "A" && ch <= "Z");
    }
    let set = new Set(banned);
    let map = new Map();
    let maxCount = 0;
    let maxWord = '';
    let word = "";
    for (let char of paragraph) {
        if (isLetter(char)) {
            word += char;
        } else if (word) {
            word = word.toLowerCase();
            if (!set.has(word)) {
                if (maxWord == '') maxWord = word.toLowerCase();
                let count = map.get(word) ? map.get(word) + 1 : 1;
                if (count > maxCount) {
                    maxCount = count;
                    maxWord = word;
                }
                map.set(word, count);
            }
            word = "";
        }
    }
    if (maxWord == '') maxWord = word.toLowerCase();
    return maxWord;
};
// @lc code=end

